src/loop_logic.rs

```diff
@@ -113,124 +140,98 @@ impl<Data> LoopHandle<Data> {
+            .register(&mut *self.inner.poll.borrow_mut(), source_token)
     }
 
-    /// Make this source update its registration
+    /// Makes this source update its registration.
     ///
     /// If after accessing the source you changed its parameters in a way that requires
     /// updating its registration.
     ///
     /// **Note**: This cannot be done from within the callback of the same source.
-    pub fn update<E: EventSource>(&self, source: &Source<E>) -> io::Result<()> {
+    pub fn update(&self, source_token: Token) -> io::Result<()> {
         self.inner
             .sources
             .borrow()
-            .get_dispatcher(source.token)
+            .get_dispatcher(source_token)
             .expect("Attempting to access a non-existent source?!")
-            .reregister(&mut *self.inner.poll.borrow_mut(), source.token)
+            .reregister(&mut *self.inner.poll.borrow_mut(), source_token)
     }
 
-    /// Disable this event source
+    /// Disables this event source.
     ///
     /// The source remains in the event loop, but it'll no longer generate events
     ///
     /// **Note**: This cannot be done from within the callback of the same source.
-    pub fn disable<E: EventSource>(&self, source: &Source<E>) -> io::Result<()> {
+    pub fn disable(&self, source_token: Token) -> io::Result<()> {
         self.inner
             .sources
             .borrow()
-            .get_dispatcher(source.token)
+            .get_dispatcher(source_token)
             .expect("Attempting to access a non-existent source?!")
             .unregister(&mut *self.inner.poll.borrow_mut())
     }
 
-    /// Remove this source from the event loop
-    ///
-    /// You are givent the `EventSource` back.
+    /// Removes this source from the event loop.
     ///
     /// **Note**: This cannot be done from within the callback of the same source.
-    pub fn remove<S: EventSource + 'static>(&self, source: Source<S>) -> S {
+    pub fn remove(&self, source_token: Token) {
         let source = self
             .inner
             .sources
             .borrow_mut()
-            .del_source(source.token)
+            .del_source(source_token)
             .expect("Attempting to remove a non-existent source?!");
-        if let Err(e) = source.unregister(&mut *self.inner.poll.borrow_mut()) {
+
+        if let Err(e) = source.unregister(&mut self.inner.poll.borrow_mut()) {
             log::warn!(
                 "[calloop] Failed to unregister source from the polling system: {:?}",
                 e
             );
         }
-        source
-            .into_source_any()
-            .downcast::<S>()
-            .map(|s| *s)
-            .unwrap_or_else(|_| panic!("Downcasting failed!"))
     }
 
-    /// Remove this event source from the event loop and drop it
+    /// Removes this source from the event loop.
     ///
-    /// The source is not given back to you but instead dropped
+    /// The source is removed but NOT unregistered.
```

Comment:
The loop dispatching logic contains code to unregister killed sources (line 270), unless this is no longer valid?

Reply:
Indeed. I removed the comment.